dev_tools/check_shellscript_set_options.py

```diff
@@ -12,22 +12,46 @@
     from pathlib import Path
 
 
-def _sets_options_or_is_nolint(line: str) -> bool:
-    return line.strip() in ["set -euxo pipefail", "# nolint(set_options)"]
+def _sets_options_or_is_nolint(line: str, expected_options: str) -> bool:
+    return line.strip() in [ expected_options, "# nolint(set_options)"]
 
 
-def _is_valid_bash_file(filename: Path) -> bool:
-    return any(_sets_options_or_is_nolint(line) for line in filename.open().readlines())
+def _is_valid_shell_file(filename: Path, expected_options: str) -> bool:
+    return any(_sets_options_or_is_nolint(line, expected_options) for line in filename.open().readlines())
+
+
+def _separate_bash_from_sh_files(filenames: Sequence[Path]) -> tuple[list[Path], list[Path]]:
+    bash_files = []
+    sh_files = []
+    for filename in filenames:
+        first_line = filename.open().readline()
```

Comment:
While I get the separation from a standpoint of code cleanliness, I think we should refactor this so that every file is only opened once.

This script is mostly IO bound and I expect python to read more than a single line when doing a `readline()`. So in the extreme case we are filling up the fs cache with all shell scripts. And then we need to evict some of those again to read those we already opened previously.

Reply:
Well spotted, I opted for readability over performance optimization. I'd say as long as you can't measure the need for a performance improvement, this is premature optimization.

Still, I'm willing to go in your direction: I could read the file content directly here and return lists of strings to only read once. This will require more memory, but only IO once. WDYT about that?

Reply:
> This will require more memory, but only IO once. WDYT about that?

Like this less then the current solution. But I agree this might be premature optimization.